### PR TITLE
Refactor parallel scan node and framework

### PR DESCRIPTION
**Pull request by yjhjstz**
Seq scan, Index scan, Indexonly scan also use a unified approach.

step 1,calc space in leader : EstimateGpParallelDSMEntrySize
step 2,Init leader DSM: InitializeGpParallelDSMEntry
step 3,Init worker DSM : InitializeGpParallelWorkers

fix #ISSUE_Number

---

### Change logs

_Describe your change clearly, including what problem is being solved or what feature is being added._

_If it has some breaking backward or forward compatibility, please clary._

### Why are the changes needed?

_Describe why the changes are necessary._

### Does this PR introduce any user-facing change?

_If yes, please clarify the previous behavior and the change this PR proposes._

### How was this patch tested?

_Please detail how the changes were tested, including manual tests and any relevant unit or integration tests._

### Contributor's Checklist

Here are some reminders and checklists before/when submitting your pull request, please check them:

- [ ] Make sure your Pull Request has a clear title and commit message. You can take [git-commit](https://github.com/cloudberrydb/cloudberrydb/blob/main/.gitmessage) template as a reference.
- [ ] Sign the Contributor License Agreement as prompted for your first-time contribution(*One-time setup*).
- [ ] Learn the [coding contribution guide](https://cloudberrydb.org/contribute/code), including our code conventions, workflow and more.
- [ ] List your communication in the [GitHub Issues](https://github.com/cloudberrydb/cloudberrydb/issues) or [Discussions](https://github.com/orgs/cloudberrydb/discussions) (if has or needed).
- [ ] Document changes.
- [ ] Add tests for the change
- [ ] Pass `make installcheck`
- [ ] Pass `make -C src/test installcheck-cbdb-parallel`
- [ ] Feel free to request `cloudberrydb/dev` team for review and approval when your PR is ready🥳

### CODE REVIEW

**Comment by avamingli** [comment]
> step 1,calc space in leader : EstimateGpParallelDSMEntrySize
step 2,Init leader DSM: InitializeGpParallelDSMEntry

Do we have a leader like PG?

**Comment by yjhjstz** [comment]
> > step 1,calc space in leader : EstimateGpParallelDSMEntrySize
> > step 2,Init leader DSM: InitializeGpParallelDSMEntry
> 
> Do we have a leader like PG?

no, logical who first enter and setup dsm, who is leader.

File: src/backend/executor/nodeIndexonlyscan.c [hunk 1]
```diff
@@ -87,33 +87,16 @@ IndexOnlyNext(IndexOnlyScanState *node)
 
 	if (scandesc == NULL)
 	{
-		if (node->ss.ps.plan->parallel_aware && estate->useMppParallelMode)
-		{
-			ParallelIndexScanDesc piscan;
-			ParallelEntryTag tag;
-			int localSliceId = LocallyExecutingSliceIndex(estate);
```
Comment (avamingli) [inline src/backend/executor/nodeIndexonlyscan.c:94]: Shall we use `localSliceId`, or remote it  when insert entry?

Comment (yjhjstz) [reply src/backend/executor/nodeIndexonlyscan.c:94]: GpFetchParallelDSMEntry has been removed , so  nothing to do with localSliceId.

Comment (avamingli) [reply src/backend/executor/nodeIndexonlyscan.c:94]: Shall we remove it when insert entry,  in GpInsertParallelDSMHash()?

Comment (avamingli) [reply src/backend/executor/nodeIndexonlyscan.c:94]: A little confused, insert entry needs a localSlicdId, but we don't use it when found entry?
How do nodes find their own dsm entry?

Comment (yjhjstz) [reply src/backend/executor/nodeIndexonlyscan.c:94]: you can found it in https://github.com/cloudberrydb/cloudberrydb/blob/main/src/backend/access/transam/parallel.c#L1898

Comment (avamingli) [reply src/backend/executor/nodeIndexonlyscan.c:94]: > No, we need to distinguish which is leader or worker.

Isn't it decided by lock `GpParallelDSMHashLock`?  Who get the lock is the first one.

Comment (yjhjstz) [reply src/backend/executor/nodeIndexonlyscan.c:94]: yes, you mentioned when remove entry, that is in GpDestroyParallelDSMEntry.

Comment (avamingli) [reply src/backend/executor/nodeIndexonlyscan.c:94]: I looked at these codes, It seems localSliceId is used to insert and destroy entry when execute plan.
We used to fetch dsm entry with localSliceId, but this pr sets parallel scandesc as other PG nodes after we insert/fetch an entry. And we use it to destroy that entry when plan ends.
The localSliceId is used to distinguish slices within a whole plan of  a session, it has nothing with leader or worker.

Do I understand right?

**Review by avamingli — APPROVED** [review]
LGTM. Thanks for your work!

File: src/backend/executor/nodeIndexonlyscan.c (hunk 1 above)
Comment (yjhjstz) [reply src/backend/executor/nodeIndexonlyscan.c:94]: right . localSliceId has nothing with leader or worker.

**Pull request merged**